### ISSUE TITLE
Docker distribution

### ISSUE THREAD

**Issue by mottosso** [getavalon/core#317]
I got an idea for how to reduce installation and update of Avalon into a single command.

```bash
$ docker run -d -p 445:445 -p 27017:27017 avalon/docker:1.0
```

Following this, database is running, dependent and cross-platform dependent files are accessible via a `samba` server, e.g. `\\192.168.99.100\avalon`

This would allow for the above command to be the entirety of the pipeline, whereby any artists working on other machines would simply connect to the share, just like how it's done locally. We could package up Python, PyQt5 and the whole shebang for every platform, and expose each as `/avalon/windows` and `/avalon/linux` etc.

With deployment this simple, it'd be trivial to include any number of hard-to-distribute dependencies and additions, such as CGWire, and updates for the end-user would take seconds and consist of altering that one command from `:1.0` to `:1.1` etc.

**Comment by tokejepsen** [comment]
Sounds very interesting!

Why ```avalon/docker:1.0``` and not just ```avalon/docker```?

**Comment by tokejepsen** [comment]
Also would we need to expose a volume for the database?

**Comment by mottosso** [comment]
> Why avalon/docker:1.0 and not just avalon/docker?

`:1.0` is for when you update and would like to know what you update from and to. You could also exclude it, which is probably a good idea for the introductory tutorials; less to worry about.

> Also would we need to expose a volume for the database?

Yes! For the tutorials, I wouldn't bother as it would store things in the container and the container could be reused during the introductory phase.

For production, it'd be a matter of adding `-v /my/local/path:/data/db` to the command. Done!

**Comment by tokejepsen** [comment]
Let get to work on this beast!

How do you see this? As a separate repository?

**Comment by tokejepsen** [comment]
Might also be nice to have a development guide, which I could imagine is just about exposing a volume to the ```avalon-setup``` repository inside of the Docker container

**Comment by mottosso** [comment]
> Let get to work on this beast!

Haha, I like your enthusiasm. :)

> How do you see this? As a separate repository?

If it's a repo, we could tell Docker hub to automatically upgrade the docker image whenever we make a GitHub release, which is quite nice. Also I'd imagine we host all binary files in Git as well, like PyQt5 for Windows, Linux and OSX, so that repo could be a good fit for that.

I was planning on having a closer look at it within the next few weeks, but if you're up for building such a Dockerfile I'd be happy to put the basics of the idea together and point at where I think it could go.

From a few experiments I made in the past few days, I reckon there are two high-level approaches.

1. Single Docker image
2. Docker Compose file

With Docker Compose, we can put it together more quickly, make it a little more reliable, and it'll scale better once deployed in production. The only downside is that we can't host docker-compose files on Docker Hub, which means one additional step for install.

```python
$ git clone ...avalon/docker
$ cd docker
$ docker-compose up
```

Which in turn adds an additional dependency; git. It also means there are one more step for updates, `git pull`.

So basically, good for us, bad for users.

The other option, a single container, would mean we need to install Samba, Mongo, Data, eventual CGWire etc into a single image, a single OS. It'd mean managing lots of dependencies and ensuring nothing conflicts. It'd mean that over time, it'd get more and more challenging, and building it would take longer and longer, grow bigger and bigger.

So basically, good for users, bad for us.

Choose! :D

**Comment by tokejepsen** [comment]
Think anyone with some tech know-how could get setup with Avalon right now, so I think this should be accommodating new users to Avalon to option 1 would be my choice.

**Comment by mottosso** [comment]
Agreed.

I've made a repo and included an example of how to achieve the effect using docker-compose, because it's simpler. What will need to happen is doing what compose does for us, which is serve (1) mongodb and (2) samba.

- https://github.com/getavalon/docker.git

I'd recommend having a look at their respective Dockerfiles and copy/pasting what fits, they likely use different OSes which may complicate things a little.

I've included Python and PyQt5 for Windows, and a "volume" directory, which is what is served on the network once the image is live. That's where we can put arbitrary things. One of the tricks here is that we'll put all files into the `volume/` directory, including other repos (ideally as submodules). That way, the Docker image itself will contain both the applications *and* the data, such that when you `docker run ...` it'll download the OS, applications and data all in one go.

The data part might make it a little awkward and non-Dockery, but what we gain is a one-line install.

**Comment by mottosso** [comment]
I noticed that you'll need to manually build the Dockerfile before running compose..

```bash
$ docker build -t avalon/docker:1.0 .
$ docker-compose up
```

You're welcome to discard the compose file as soon as you've gotten out of it what you need.

**Comment by mottosso** [comment]
For completeness, I had one more thought that could make the install and use even simpler, which is that rather than bundle Python for Windows, Linux and OSX in order to run standalone software like the Launcher, we could *stream* the GUI from the Docker image and into the native window server of Linux and to clients for Windows and OSX like [VcXsrv](https://sourceforge.net/projects/vcxsrv/).

Streaming is generally slower; it's basically a VNC connection per window, and having experimented with it on Windows, it is rather sluggish. But if there's something to do about that, then we'd be able to reap the same flexibility as hosting software which is later viewed from a browser, like CGWire. It don't need to bundle Chrome for every platform, and that's really powerful.

Alternatively, we could have a look at making the launcher into a web app so we could benefit from the browser too.

**Comment by tokejepsen** [comment]
That sounds cool. Maybe a 1.5 release :)

**Comment by mottosso** [comment]
Yes. :)

Could be both. Could distribute files, and keep a port open for streaming. They'd be using the same source so could be used in parallel and where relevant. You can stream a GUI to a different machine too, and possibly into the browser.

**Comment by tokejepsen** [comment]
I think we can close this issue, since we are working on it atm?
